File: js/godot/resource.js
import {getString, Variant} from "./variant";
import {ResourceType} from "./type";
import {OggSampleExtension} from "./extensions/oggSampleExtension";
import {WavSampleExtension} from "./extensions/wavSampleExtension";
import {CompressedTexture, StreamTexture} from "./types/texture";
import {BinaryReader, SEEK_CUR} from "../binaryReader";
import {Mp3SampleExtension} from "./extensions/mp3SampleExtension";
import {DataDecompressor} from "./compressedData";
import {FontFileExtension} from "./extensions/fontFileExtension";

const TYPE_RESOURCE = 0;
const TYPE_BINARY = 1;

export class Resource extends ResourceType {
  constructor(reader) {
    super();
    this.reader = reader;
  }

  async load() {
    let magic = this.reader.readChars(4);
    if (magic === 'RSCC') {
      const data = await new DataDecompressor().decompress(this.reader);
      this.reader = new BinaryReader(data, 'little');
    } else if (magic !== 'RSRC') {
      this.type = TYPE_BINARY;
      this.reader.seek(-4, SEEK_CUR);
      let parser = this.getParserForMagic(magic);
      if (parser) {
        this.parser = new parser(this.reader);
      }
      return this;
    }
    this.type = TYPE_RESOURCE;
    this.reader.endian = (!!this.reader.readInt32()) ? 'big' : 'little';  // is big-endian
    this.useReal64 = !!this.reader.readInt32();
    this.verMajor = this.reader.readInt32();
    this.verMinor = this.reader.readInt32();
    this.formatVersion = this.reader.readInt32();
    this.typeName = this.reader.readString();
    this.importOffset = this.reader.readInt64();
    this.flags = this.reader.readInt32();
    this.uid = this.reader.readInt64();
    this.reader.read(44);  // reserved
    this.stringTable = this.reader.readArrayT(this.reader.readString.bind(this.reader), this.reader.readInt32());
    this.externalResources = this.reader.readArrayT(() => {
      return {
        typeName: this.reader.readString(),
        path: this.reader.readString(),
        uid: ((this.flags & 2) !== 0) ? this.reader.readInt64() : 0
      }
    }, this.reader.readInt32());
    this.internalResources = this.reader.readArrayT(() => {
      return {
        path: this.reader.readString(),
        offset: this.reader.readInt64()
      }
    }, this.reader.readInt32());
    let typeName = this.reader.readString();
    let properties = {};
    let numProperties = this.reader.readInt32();
    for (let i = 0; i < numProperties; i++) {
      properties[getString(this.reader, this.stringTable)] = new Variant(this.reader).value;
    }
    this.resource = {
      typeName,
      properties
    }
  }

  getParserForMagic(magic) {
    switch (magic) {
      case 'GDST':
        return StreamTexture;
      case 'GST2':
        return CompressedTexture;
      default:
        return null;
    }
  }

  getExtension() {
    switch (this.typeName) {
      case 'AudioStreamOGGVorbis':
        return new OggSampleExtension()
      case 'AudioStreamSample':
        return new WavSampleExtension()
      case 'AudioStreamMP3':
        return new Mp3SampleExtension()
      case 'FontFile':
        return new FontFileExtension()
      default:
        return null;
    }
  }

  async createPreview() {
    if (this.type === TYPE_BINARY) {
      if (this.parser && typeof this.parser.createPreview == 'function') {
        return await this.parser.createPreview();
      } else {
        return document.createElement('div');
      }
    }
    const ext = this.getExtension();
    if (ext) {
      return ext.createPreview(this);
    } else {
      return document.createElement('div');
    }
  }

  getExport() {
    return this.getExtension().exportFile(this);
  }
}

File: src/godot/resource.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex, MutexGuard};
use anyhow::anyhow;
use bytes::{Buf, Bytes};
use crate::base::asset::{Asset, AssetMetadata, UnsupportedAsset};
use crate::base::asset::bundle::BundleFile;
use crate::base::asset::provider::AssetProvider;
use crate::base::asset::types::AssetType;
use crate::godot::variant::{get_string, Variant};
use crate::godot::wrappers::audio::audio_stream_mp3::Mp3StreamWrapper;
use crate::godot::wrappers::audio::audio_stream_ogg::OggStreamWrapper;
use crate::godot::wrappers::audio::audio_stream_wav::WavStreamWrapper;
use crate::logger::{info, warning};

#[derive(Debug, Clone)]
pub(crate) struct ExternalResourceReference {
    type_name: String,
    path: String,
    uid: i64
}

#[derive(Debug, Clone)]
pub(crate) struct InternalResourceReference {
    path: String,
    offset: usize
}

#[derive(Debug, Clone)]
pub(crate) struct Resource {
    pub(crate) r#type: String,
    pub(crate) properties: HashMap<String, Variant>
}

#[derive(Debug, Clone)]
pub(crate) struct ResourceFile {
    use_real_64: bool,
    major: i32,
    minor: i32,
    format_version: i32,
    type_name: String,
    import_offset: usize,
    flags: i32,
    uid: i64,
    string_table: Vec<String>,
    external_resources: Vec<ExternalResourceReference>,
    internal_resources: Vec<InternalResourceReference>,
    resource: Resource
}

impl ResourceFile {
    pub(crate) fn from_bytes(data: &mut Bytes) -> anyhow::Result<Self> {
        // TODO compressed (RSCC)
        if data.get_chars(4) != "RSRC" {
            return Err(anyhow!("not a resource file"));
        }

        let little_endian = data.get_i32_le() == 0;
        let use_real_64 = data.get_i32_ordered(little_endian) != 0;
        let major = data.get_i32_ordered(little_endian);
        let minor = data.get_i32_ordered(little_endian);
        let format_version = data.get_i32_ordered(little_endian);
        let type_name = data.get_string_ordered(little_endian);
        let import_offset = data.get_i64_ordered(little_endian) as usize;
        let flags = data.get_i32_ordered(little_endian);
        let uid = data.get_i64_ordered(little_endian);
        data.advance(44);  // reserved
        let mut string_table = Vec::new();
        let num_strings = data.get_i32_ordered(little_endian);
        for _ in 0..num_strings {
            string_table.push(data.get_string_ordered(little_endian));
        }
        let mut external_resources = Vec::new();
        for _ in 0..data.get_i32_ordered(little_endian) {
            external_resources.push(ExternalResourceReference {
                type_name: data.get_string_ordered(little_endian),
                path: data.get_string_ordered(little_endian),
                uid: if flags & 2 != 0 { data.get_i64_ordered(little_endian) } else { 0 },
            });
        }
        let mut internal_resources = Vec::new();
        for _ in 0..data.get_i32_ordered(little_endian) {
            internal_resources.push(InternalResourceReference {
                path: data.get_string_ordered(little_endian),
                offset: data.get_i64_ordered(little_endian) as usize,
            })
        }
        let resource_type = data.get_string_ordered(little_endian);
        let mut properties = HashMap::new();
        for _ in 0..data.get_i32_ordered(little_endian) {
            let key = get_string(data, &string_table, true).unwrap();
            let val = Variant::from_bytes(data, &string_table, format_version < 3, true, major)?;
            properties.insert(key, val);
        }
        // TODO: this will fail if there is more than one resource
        // if data.get_chars(4) != "RSRC" {
        //     return Err(anyhow!("resource file was not read properly"));
        // }
        Ok(ResourceFile {
            use_real_64,
            major,
            minor,
            format_version,
            type_name,
            import_offset,
            flags,
            uid,
            string_table,
            external_resources,
            internal_resources,
            resource: Resource {
                r#type: resource_type,
                properties,
            },
        })
    }
}

impl AssetProvider for ResourceFile {
    fn list_assets(&self) -> Vec<AssetMetadata> {
        // TODO: technically, there can be more than one resource per file
        //  Right now we just read the first one
        vec![AssetMetadata {
            name: "resource".to_string(),
            asset_type: AssetType::Misc,
            id: "owo whats this".to_string(),
        }]
    }

    fn get_asset(&self, id: String, parent: Option<&mut MutexGuard<Box<dyn BundleFile + Send>>>) -> Option<Arc<Mutex<Box<dyn Asset>>>> {
        Some(Arc::new(Mutex::new(match self.type_name.as_str() {
            "AudioStreamOGGVorbis" => Box::new(OggStreamWrapper::wrap(&self.resource).unwrap()) as Box<dyn Asset>,
            "AudioStreamSample" => Box::new(WavStreamWrapper::wrap(&self.resource).unwrap()) as Box<dyn Asset>,
            "AudioStreamMP3" => Box::new(Mp3StreamWrapper::wrap(&self.resource).unwrap()) as Box<dyn Asset>,
            other => {
                warning!("unknown type {other}");
                info!("{:#?}", self);
                Box::new(UnsupportedAsset {}) as Box<dyn Asset>
            }
        })))
    }
}
